fault/verilator_target.py

```diff
@@ -1,219 +1,99 @@
-        std::cerr << "top->{name} = " << test[{i}].value << ", ";
-'''
-                i += 1
-        else:
-            if isinstance(port, m.ArrayType) and \
-                    not isinstance(port.T, m._BitType):
-                for _name in flattened_names(port):
-                    output_str += f'''\
-        check(\"{name}{_name}\", top->{name}{_name}, test[{i}], i);
-'''
-                    i += 1
-            else:
-                output_str += f'''\
-        check(\"{name}\", top->{name}, test[{i}], i);
-'''
-                i += 1
-
-    source += f'''\
-        std::cerr << std::endl;
-        std::cerr << "Checking Outputs: ";
-{output_str}
-        std::cerr << std::endl;
-        top->eval();
-        std::cerr << "{"="*80}" << std::endl;
-'''
-    source += '''\
-    }
-'''
-
-    source += '''
-    delete top;
-    std::cout << "Success" << std::endl;
-    exit(0);
-}'''
-
-    return source
-
-
-def compile_verilator_harness(basename, circuit, tests, input_ranges=None):
-    if config.get_compile_dir() == 'callee_file_dir':
-        (_, filename, _, _, _, _) = \
-            inspect.getouterframes(inspect.currentframe())[1]
-        file_path = os.path.dirname(filename)
-        filename = os.path.join(file_path, basename)
-    else:
-        filename = basename
-
-    if callable(tests):
-        tests = generate_function_test_vectors(circuit, tests, input_ranges)
-    verilatorcpp = harness(circuit, tests)
-
-    with open(filename, "w") as f:
-        f.write(verilatorcpp)
-
-
-def run_verilator_test(verilog_file_name, driver_name, top_module,
-                       verilator_flags="", build_dir="build"):
-    if isinstance(verilator_flags, list):
-        if not all(isinstance(flag, str) for flag in verilator_flags):
-            raise ValueError("verilator_flags should be a str or list of strs")
-        verilator_flags = " ".join(verilator_flags)
-    assert not subprocess.call(
-        f'verilator -Wall -Wno-INCABSPATH -Wno-DECLFILENAME {verilator_flags}'
-        f' --cc {verilog_file_name}.v --exe {driver_name}.cpp'
-        f' --top-module {top_module}',
-        cwd=build_dir, shell=True
-    )
-    assert not subprocess.call(
-        f'make -C obj_dir -j -f V{top_module}.mk V{top_module}', cwd=build_dir,
-        shell=True)
-    assert not subprocess.call('./obj_dir/V{}'.format(top_module),
-                               cwd=build_dir, shell=True)
+}}
+"""
 
 
 class VerilatorTarget(Target):
-    def __init__(self, circuit, test_vectors, directory="build/", flags=[]):
-        super().__init__(circuit, test_vectors)
-        self._directory = directory
-        self._flags = flags
+    def __init__(self, circuit, actions, directory="build/", flags=[]):
+        super().__init__(circuit, actions)
+        self.directory = Path(directory)
+        self.flags = flags
+
+    @staticmethod
+    def generate_action_code(i, action):
+        if isinstance(action, actions.Poke):
```

Comment:
need to deal with nested arrays --> flatten names

Reply:
Yes, the array flattening stuff is quite annoying, I'm not sure if there's a less complex solution to this.

Reply:
Done.